rfcs/0026-memory-control.md

```diff
@@ -0,0 +1,113 @@
+
+## Motivation
+
+We had discussed a lot about how to allocate and manage memory before. [RFC: Yet another simple idea for memory management](https://singularity-data.quip.com/CldAAcFmzZSO/Yet-another-simple-idea-for-memory-management) proposed a simple way to control the memory usage of streaming executors by controlling the *watermark epoch* of the LRU caches on a compute node, which was finally adopted in our system, surpassing [RFC: Simpler Dynamic Memory Management Among LRU-Caches](https://singularity-data.quip.com/A1kHAOBUo3Im/RFC-Simpler-Dynamic-Memory-Management-Among-LRU-Caches) and [RFC: Dynamic Memory Budget of Streaming Operators](https://singularity-data.quip.com/J9KYAQc2xIbr/RFC-Dynamic-Memory-Budget-of-Streaming-Operators).
+
+Recently, [RFC: Coarse grained resource management for batch query engine](https://github.com/risingwavelabs/rfcs/pull/11) tried to control the memory of batch executors as well, but the 1st and 2nd version of the design was controversial because some designs were too complicated or infeasible. Luckily, the RFC has been refined a lot during the discussion, which finally comes to the solution of this RFC. 
+
+## Goals
+
+- The total memory usage is controlled to prevent OOM (Out-of-Memory).
+- The streaming and batch executors should be isolated as far as possible.
+
+## Design
+
+Memory of compute node is a scarce resource. Due to the nature of streaming processing, the memory will be sooner or later consumed out. Therefore, we need to limit the memory used by each component to prevent OOM.
+
+![](images/overall-memory-allocation.drawio.svg)
+
+**This RFC will focus on the memory controlling on both streaming and batch executors**, assuming the storage, network buffer, system metadata, etc. always take a fixed size of memory.
+
+Our approach uses one background coroutine to continuously run the loop below (inspired by @yuhao-su):
+
+1. **Measure**: Get the current memory usage
+2. **Policy**: Decide whether and which component needs to release memory
+3. **Action**: Take actions to release memory
+
+### Measure: How to measure memory usage?
+
+Here we propose to use the simple and stupid way to calculate the memory usage - **estimating** the memory consumed by all the in-memory data structures and aggregating in levels. This is opposite to the solutions that rely on memory allocators to get an exact memory usage.
+
+How to estimate memory usage depends on the data structures:
+
+- For indexing structures like `HashMap`, `BTreeMap`, etc., we could estimate the memory size according to the `capacity()` or something else. 
+   - @liurenjie1024: It also seems possible to inject an `Allocator` to count the exact allocated space, as long as it exposed such generic type parameter.
+- For pure data structures like `DataChunk`, `Vec<Datum>`, etc., we could compute the exact size it used by summing up the sizes of internal arrays & variables as well as padding and other overhead.
```

Comment:
It seems that injecting `Allocator` to `DataChunk` is easier to implement, otherwise we need to override several collections with contains `DataChunk` such as `Vec`, `HeapElement`

Reply:
It will looks similar to https://github.com/risingwavelabs/risingwave/pull/5932

Reply:
> It seems that injecting `Allocator` to `DataChunk` is easier to implement, otherwise we need to override several collections with contains `DataChunk` such as `Vec`, `HeapElement`

Easy but meaningless.

Reply:
The complexity is that we need to override container methods, you can think about how to estimate the size of `Vec<DataChunk>`. Anyway it's not that clean.